src/v2/manifest/mod.rs

```diff
@@ -19,62 +19,68 @@ impl Client {
-                    error!("{}", msg);
-                    return Box::new(futures::future::err(Error::from(msg)));
+                    return Box::new(futures::future::err::<_, _>(Error::from(format!(
+                        "failed to parse url from string '{}': {}",
+                        ep, e
+                    ))));
                 }
-            };
-            let mtype = mediatypes::MediaTypes::ManifestV2S2.to_string();
-            req.headers_mut().append(
-                header::ACCEPT,
-                match header::HeaderValue::from_str(&mtype) {
-                    Ok(headervalue) => headervalue,
-                    Err(e) => {
-                        let msg = format!("failed to parse HeaderValue from str: {}:", e);
-                        error!("{}", msg);
-                        return Box::new(futures::future::err::<_, _>(Error::from(msg)));
-                    }
-                },
-            );
-            req
+            }
         };
-        let freq = self.hclient.request(req);
-        let fres = freq
-            .from_err()
-            .inspect(move |_| {
-                trace!("GET {:?}", url);
-            })
-            .and_then(|r| {
-                let status = r.status();
+
+        let fres = self
+            .build_reqwest(reqwest::async::Client::new().get(url))
+            .send()
+            .map_err(|e| ::errors::Error::from(format!("{}", e)))
+            .and_then(|res| {
+                trace!("GET '{}' status: {:?}", res.url(), res.status());
+                let status = res.status();
                 trace!("Got status: {:?}", status);
                 match status {
-                    hyper::StatusCode::OK => Ok(r),
-                    _ => Err(format!("get_manifest: wrong HTTP status '{}'", status).into()),
+                    reqwest::StatusCode::OK => Ok(res),
+                    _ => Err(format!("GET {}: wrong HTTP status '{}'", res.url(), status).into()),
                 }
             })
-            .and_then(|r| {
-                r.into_body().concat2().map_err(|e| {
-                    format!("get_manifest: failed to fetch the whole body: {}", e).into()
-                })
+            .and_then(|mut res| {
+                std::mem::replace(res.body_mut(), reqwest::async::Decoder::empty())
+                    .concat2()
+                    .map_err(|e| ::errors::Error::from(format!("{}", e)))
+                    .join(futures::future::ok(res))
             })
-            .and_then(|body| Ok(body.into_bytes().to_vec()));
+            .map_err(|e| ::errors::Error::from(format!("{}", e)))
+            .and_then(|(body, res)| {
+                let body_vec = body.to_vec();
+                let len = body_vec.len();
+                let status = res.status();
+                Ok((
+                    body_vec,
+                    res.headers()
+                        .get("docker-content-digest")
```

Comment:
This seems to be with uppercase initials in registry API specs: https://docs.docker.com/registry/spec/api/#digest-header

Reply:
quay doesn't follow that apparently:

```console
curl -v --get https://quay.io/v2/coreos/etcd/manifests/latest
(...)
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 200 
< server: nginx/1.14.2
< date: Thu, 24 Jan 2019 17:49:22 GMT
< content-type: application/vnd.docker.distribution.manifest.v1+prettyjws
< content-length: 6655
< docker-content-digest: sha256:5b6691b7225a3f77a5a919a81261bbfb31283804418e187f7116a0a9ef65d21d
< x-frame-options: DENY
< strict-transport-security: max-age=63072000; preload
(...) 
```

Reply:
Ah, they are case-insensitive according to https://tools.ietf.org/html/rfc7230#section-3.2. It would be nice to use the spec format if the getter supports this internally. Otherwise don't mind.

Reply:
I verified that the current code is case-insensitive

Reply:
@steveeJ you didn't change the header name constant though. Was that on purpose or you just forgot to?

Reply:
I left it lower-case cause I thought that makes more sense from a perspective of being case-insensitive. At least that's the config I'm using in my code editors: case-insensitive if all-lowercase.